src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
-        return null;
+        checkIndexPosition(index);
+        return unlinkNode(getNodeByIndex(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> current = first; current != null; current = current.next) {
+                if (current.value == null) {
+                    unlinkNode(current);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> current = first; current != null; current = current.next) {
+                if (object.equals(current.value)) {
+                    unlinkNode(current);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == ZERO_SIZE;
+    }
+
+    private void checkIndexElement(int index) {
+        if (!isIndexElement(index)) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private void checkIndexPosition(int index) {
+        if (!isIndexPosition(index)) {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    private boolean isIndexElement(int index) {
+        return index > MINUS_ONE && index <= size;
+    }
+
+    private boolean isIndexPosition(int index) {
+        return index > MINUS_ONE && index < size;
+    }
+
+    private T unlinkNode(Node<T> targetNode) {
+        T value = targetNode.value;
+        Node<T> previousNode = targetNode.previous;
+        Node<T> nextNode = targetNode.next;
+        if (previousNode == null) {
+            first = nextNode;
+        } else {
+            previousNode.next = nextNode;
+            targetNode.previous = null;
+        }
+        if (nextNode == null) {
+            last = previousNode;
+        } else {
+            nextNode.previous = previousNode;
+            targetNode.next = null;
+        }
+        targetNode.value = null;
+        size--;
+        return value;
+    }
+
+    private void linkPreviousElement(T value, Node<T> targetNode) {
+        Node<T> previousNode = targetNode.previous;
+        Node<T> newNode = new Node<>(value,previousNode,targetNode);
+        targetNode.previous = newNode;
+        if (previousNode == null) {
+            first = newNode;
+        } else {
+            previousNode.next = newNode;
+        }
+        size++;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index < (size >> SIZE_BIT_DIVIDER)) {
+            Node<T> iterateNodeFromFirst = first;
+            for (int i = 0; i < index; i++) {
+                iterateNodeFromFirst = iterateNodeFromFirst.next;
+            }
+            return iterateNodeFromFirst;
+        } else {
+            Node<T> iterateNodeFromLast = last;
+            for (int i = size - SIZE_SHIFTER; i > index; i--) {
```

Comment:
just use / 2, will be more readable than bitwise operations with constants

Reply:
ok, I'll correct it. I always thougth that it is more prioritized to use bitwise operation. I.e. should I try to avoid it? Or should use it appropriately ?

Reply:
IMHO readability should take precedence over a negligible performance gain, as maintainability and clarity of code are generally more important